hook.sh

```diff
@@ -5,79 +5,102 @@
-    DEDYNAUTH=`pwd`./dedynauth
+if [ -z "$DEDYN_TOKEN" ]; then
+    DEDYNAUTH=$(pwd)./dedynauth
     >&2 echo "Variable \$DEDYN_TOKEN not found. Please set DEDYN_TOKEN=(your dedyn.io token) to your dedyn.io access token in $DEDYNAUTH, e.g."
-    >&2 echo 
+    >&2 echo ""
     >&2 echo "DEDYN_TOKEN=d41d8cd98f00b204e9800998ecf8427e"
     exit 2
 fi
 
-if [ -z "$DEDYN_NAME" ] ; then
-    DEDYNAUTH=`pwd`./dedynauth
+if [ -z "$DEDYN_NAME" ]; then
+    DEDYNAUTH=$(pwd)./dedynauth
     >&2 echo "Variable \$DEDYN_NAME not found. Please set DEDYN_NAME=(your dedyn.io name) to your dedyn.io name in $DEDYNAUTH, e.g."
-    >&2 echo
+    >&2 echo ""
     >&2 echo "DEDYN_NAME=foobar.dedyn.io"
     exit 3
 fi
 
-if [ -z "$CERTBOT_DOMAIN" ] ; then
-    DEDYNAUTH=`pwd`./dedynauth
+if [ -z "$CERTBOT_DOMAIN" ]; then
+    DEDYNAUTH=$(pwd)./dedynauth
     >&2 echo "It appears that you are not running this script through certbot (\$CERTBOT_DOMAIN is unset). Please call with: certbot --manual-auth-hook=$0"
     exit 4
 fi
 
-if [[ ! $(type -P curl) ]] ; then
+if [ ! "$(type -P curl)" ]; then
     >&2 echo "Please install curl to use certbot with dedyn.io."
     exit 5
 fi
 
->&2 echo "Setting challenge to ${CERTBOT_VALIDATION} ..."
+echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
 
 # Figure out subdomain infix by removing zone name and trailing dot
 # foobar.dedyn.io gives "" while a.foobar.dedyn.io gives ".a"
 domain=.$CERTBOT_DOMAIN
 infix=${domain%.$DEDYN_NAME}
 
+# Remove leading wildcard from infix, if present
+# *.foobar.dedyn.io gives "" while *.a.foobar.dedyn.io gives ".a"
+infix=${infix#.\*}
+
 args=( \
-    '-Ss' \
+    '-sSLf' \
     '-H' "Authorization: Token $DEDYN_TOKEN" \
     '-H' 'Accept: application/json' \
     '-H' 'Content-Type: application/json' \
-    '-d' '[{"subname":"_acme-challenge'"$infix"'", "type":"TXT", "records":["\"'"$CERTBOT_VALIDATION"'\""], "ttl":60}]' \
-    '-o' '/dev/null' \
 )
 
-# set ACME challenge (overwrite if possible, create otherwise)
-curl -X PUT "${args[@]}" -f "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/"
+# Letsencrypt uses a TTL of 60 seconds on their resolvers. If we're getting
+# wildcard certificates (e.g. foobar.dedyn.io and *.foobar.dedyn.io), multiple
+# certificates for the same domain (e.g. RSA and ECDSA), or both, we'll need to
+# wait 60 seconds for the records to expire before verification begins. We'll
+# store any existing challenges and publish them along with the current one.
+# Set timeout_ttl=yes in ./dedynauth if you want to enable this.
+# You'll probably also want to use the optional cleanup hook to keep records
+# from accumulating.
 
->&2 echo "Verifying challenge is set correctly. This can take up to 2 minutes."
->&2 echo "Current Time: `date`"
+readarray -t acme_live <<< "$(curl "${args[@]}" -X GET "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/_acme-challenge$infix/TXT" \
+    | grep -o '"records":\[[^]]*]' | grep -o '\\"[^,]*\\"')"
```

Comment:
I have several comments here:

- The URL is missing the final slash. This causes an unnecessary redirect.
- You can use `readarray -t acme_live < <(curl ...)` to simplify and avoid the unnecessary string parsing.
- The code make assumptions about the JSON formatting. Arbitrary whitespace is allowed before and and after the colon `:`. To take this into account, I suggest piping the `curl` output through `tr -d '\n'` and replacing the `:` in the regex by `[[:space:]]*:[[:space:]]*`.
- Records are currently split by `,` and later reassembled. First of all, this breaks in case a challenge contains a comma `,`. Second, I don't think it is necessary at all to split the records. You can just `grep -o '"\".*\""'` to obtain the whole list of JSON strings (without the enclosing brackets). Later, you can just add the current challenge if it is not contained in this string.
- Also, this command writes `curl: (22) The requested URL returned error: 404` to stderr if the RRset does not yet exist, so the user will be confronted with an error even though everything is fine. It would be better to handle this case gracefully. The easiest way to achieve this may be to query `rrsets/$DEDYN_NAME/rrsets/?subname=_acme-challenge$infix&type=TXT`, which returns a JSON array with zero or one elements. If the RRset is not found, the return value will be `[]`, but the exit code will be a success to. The regular expression should work nevertheless.

Reply:
Great ideas all around! Reworking this to read the whole records array into an ordinary variable and appending the challenge (if it doesn't already exist) saves us an array and a loop.

Great to know about the final slash/redirect. So many future headaches avoided :-)